KAFKA-5886: Introduce delivery.timeout.ms producer config (KIP-91)

yuyang08 opened the pull request:
This change is based on @sutambe 's change https://github.com/apache/kafka/pull/3849 earlier. 

primary changes in this pr:

1. In RecordAccumulator.java, use `inFlightBatches` to track the in-flight batches,  instead of using `soonToExpireInFlightsBatches` to only track the soon-to-expire batches.  With this change, in RecordAccumulator.expiredBatches, we check both `inFlightBatches` and `batches` to find the expired batches. 

2. Fixed the test failures in SenderTest.java and RecordAccumulatorTest.java.

sutambe committed:
KAFKA-5886: Introduce delivery.timeout.ms producer config

Commit:
KAFKA-5886: update kafka pull request 3849, and fix SenderTest and RecordAccumulatorTest

yuyang08 commented:
@apurvam, @becketqin, @guozhangwang, @ijuma   could you help to review this change?

guozhangwang commented:
Thanks @yuyang08 , we will take a look at this PR asap.

yuyang08 commented:
@guozhangwang , @ijuma , @apurvam , @becketqin  friendly ping ... your feedback will help us to iterate faster

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -458,6 +462,25 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
+        // when the sum of lingerMs and requestTimeoutMs overflows a long we make sure deliveryTimeoutMs is at least
+        // equal to lingerMs.
+        boolean overflow = lingerMs + requestTimeoutMs < 0L;
+        boolean invalid = overflow ? deliveryTimeoutMs < lingerMs : deliveryTimeoutMs < lingerMs + requestTimeoutMs;
```
This check is a little weird to me. It seems arbitrary to check only `lingerMs` in the case of overflow. Wouldn't it make more sense to use `Integer.MAX_VALUE`?

yuyang08 replied:
updated the sanity checking logic.  changed  `lingerMs` and `deliveryTimeoutMs` to integer type. with that, we convert them to long for addition and do not need to worry about the overflow.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -458,6 +462,25 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
     }
 
+    private static long configureDeliveryTimeout(ProducerConfig config) {
+        long deliveryTimeoutMs = config.getLong(ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG);
```
I'm wondering if we should set this value automatically if the value is not overridden. It would be annoying to get this config error after an upgrade if I have overridden the request timeout. Maybe we could use the max of the default delivery timeout and the sum of `requestTimeoutMs` and `lingerMs`?

yuyang08 replied:
my concern is that setting the value automatically may hide some issues from the user, and it also requires more explanation. It might be better to keep it simple. The issue that the users can run into an upgrade is that they override `requestTimeoutMs` with a large value, and that violates the required variant `deliveryTimeout > requestTimeoutMs + lingerMs`. In that case, the user will be notify the error during producer initialization, and it shall not take much effort to fix it.

hachikuji replied:
Yeah, I appreciate the concern. However, there is already some precedent for redefining defaults based on provided configurations. For example, when idempotence is enabled, we override `retries` if it has not been explicitly provided by the user. We log an info message so that the user knows the default has been adjusted. 

In general, I think we should try to avoid compatibility issues even in configuration as long as it is safe to do so. By using a larger request timeout, the user has already declared willingness to await for the request timeout to receive a produce acknowledgement, so redefining the default delivery timeout seems reasonable and saves an annoying config update.

ijuma replied:
A reason for the error is that a longer request timeout is not unlikely to have been done as a workaround for the issue that delivery timeout is fixing. So, it might provide the user with an opportunity to fix that. A warning might be a less heavy handed way to achieve this though.

yuyang08 replied:
updated the change to use `requestTimeoutMs + lingerMs` if it is larger than `deliveryTimeoutMs` setting, and log a warning message.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -99,6 +99,12 @@
+    public static final String DELIVERY_TIMEOUT_MS_CONFIG = "delivery.timeout.ms";
+    private static final String DELIVERY_TIMEOUT_MS_DOC = "An upper bound on the time to report success or failure after Producer.send() returns. "
+                                                          + "Producer may report failure to send a message earlier than this config if all the retries are exhausted or "
+                                                          + "a record is added to a batch nearing expiration.";
```
The latter half here may not be very clear to users since it's a bit low level. Is there actually a lower bound for the timeout or can it be arbitrarily small depending on how long the batch remains open?

yuyang08 replied:
updated the doc string on the lower bound explanation.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -224,7 +230,7 @@
         CONFIG = new ConfigDef().define(BOOTSTRAP_SERVERS_CONFIG, Type.LIST, Collections.emptyList(), new ConfigDef.NonNullValidator(), Importance.HIGH, CommonClientConfigs.BOOTSTRAP_SERVERS_DOC)
                                 .define(BUFFER_MEMORY_CONFIG, Type.LONG, 32 * 1024 * 1024L, atLeast(0L), Importance.HIGH, BUFFER_MEMORY_DOC)
-                                .define(RETRIES_CONFIG, Type.INT, 0, between(0, Integer.MAX_VALUE), Importance.HIGH, RETRIES_DOC)
+                                .define(RETRIES_CONFIG, Type.INT, Integer.MAX_VALUE, between(0, Integer.MAX_VALUE), Importance.HIGH, RETRIES_DOC)
```
Can you mention this default change in the upgrade notes?

yuyang08 replied:
updated the upgrade doc on this.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -158,34 +158,50 @@ public void abort(RuntimeException exception) {
+        if (this.finalState.get() != FinalState.SUCCEEDED) {
+            if (tryFinalState == FinalState.SUCCEEDED) {
+                // Log if a previously unsuccessful batch succeeded later on.
+                log.debug("ProduceResponse returned {} for {} after batch had already been {}.", tryFinalState,
```
Maybe helpful to mention the baseOffset in this message?

yuyang08 replied:
added `baseOffset` info in the log message

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -158,34 +158,50 @@ public void abort(RuntimeException exception) {
+                log.debug("ProduceResponse returned {} for {} after batch had already been {}.", tryFinalState,
+                    topicPartition, this.finalState.get());
             }
+            // FAILED --> ABORTED and ABORTED --> FAILED transitions are ignored.
```
Since this should not be a common case, I'm wondering if we should just log the outcome in all cases where we cancelled early. It may help with debugging timeout issues.

yuyang08 replied:
added logging here

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -253,7 +262,7 @@ private MemoryRecordsBuilder recordsBuilder(ByteBuffer buffer, byte maxUsableMag
      */
     private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, Header[] headers,
-                                         Callback callback, Deque<ProducerBatch> deque) {
+        Callback callback, Deque<ProducerBatch> deque) {
```
nit: this seems unintentional

yuyang08 replied:
restored the space

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -234,6 +240,7 @@
                                 .define(COMPRESSION_TYPE_CONFIG, Type.STRING, "none", Importance.HIGH, COMPRESSION_TYPE_DOC)
                                 .define(BATCH_SIZE_CONFIG, Type.INT, 16384, atLeast(0), Importance.MEDIUM, BATCH_SIZE_DOC)
                                 .define(LINGER_MS_CONFIG, Type.LONG, 0, atLeast(0L), Importance.MEDIUM, LINGER_MS_DOC)
+                                .define(DELIVERY_TIMEOUT_MS_CONFIG, Type.LONG, 120 * 1000, atLeast(0L), Importance.MEDIUM, DELIVERY_TIMEOUT_MS_DOC)
```
Does delivery timeout need to be a long? I think `Integer.MAX_VALUE` is about 24 days. Not sure anything beyond that is really useful in practice.

Obviously it's a bit silly that we use a long for `linger.ms` as well.

yuyang08 replied:
good point. updated `deliveryTimeoutMs` to `int`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,37 +281,55 @@ private boolean isMuted(TopicPartition tp, long now) {
+    private void maybeUpdateNextBatchExpiryTime(ProducerBatch batch) {
+        // We determine the earliest time any batch may expire. We use
+        // this value later on to determine the max time we can sleep in poll.
+        // We assume that the batch at the front of the deque will always be the next to expire.
```
This comment does not seem very relevant in this context.

yuyang08 replied:
removed the comment

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,37 +281,55 @@ private boolean isMuted(TopicPartition tp, long now) {
+                    }
+                }
+            }
+            List<ProducerBatch> mayExpireBatches = inFlightBatches.get(entry.getKey());
```
nit: this could be named more clearly. Maybe `partitionInFlightBatches`?

yuyang08 replied:
good point. updated it to `partitionInFlightBatches`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,37 +281,55 @@ private boolean isMuted(TopicPartition tp, long now) {
+            // Expire in-flight batches in the order of draining. Expire if the final state of the
+            // batch is not known by (batch's creation time + deliveryTimeoutMs).
+
             TopicPartition tp = entry.getKey();
```
nit: this is unused

yuyang08 replied:
removed

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,37 +281,55 @@ private boolean isMuted(TopicPartition tp, long now) {
         List<ProducerBatch> expiredBatches = new ArrayList<>();
         for (Map.Entry<TopicPartition, Deque<ProducerBatch>> entry : this.batches.entrySet()) {
-            Deque<ProducerBatch> dq = entry.getValue();
+            // Expire in-flight batches in the order of draining. Expire if the final state of the
```
Can we explain in this comment why we use this order? Intuitively, I would expect that we'd expire the oldest stuff first so that the callbacks are invoked in the order of sending.

yuyang08 replied:
this is a comment that was in https://github.com/apache/kafka/pull/3849. updated the comment.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -311,16 +338,27 @@ private boolean isMuted(TopicPartition tp, long now) {
-                insertInSequenceOrder(deque, batch);
-            else
-                deque.addFirst(batch);
+        if (!batch.hasReachedDeliveryTimeout(deliveryTimeoutMs, now)) {
```
Hmm.. Shouldn't we be expiring the batch in the `else` case? Otherwise it seems like we may lose track of the batch.

yuyang08 replied:
good catch. expire the batch in `else` branch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -311,16 +338,27 @@ private boolean isMuted(TopicPartition tp, long now) {
-                deque.addFirst(batch);
+        if (!batch.hasReachedDeliveryTimeout(deliveryTimeoutMs, now)) {
+            batch.reenqueued(now);
+            maybeRemoveFromInflightBatches(batch);
```
Unless I'm missing something, we only remove from the `inFlightBatches` collection when we reenqueue and when we encounter a delivery timeout. Don't we need to remove on completion or failure as well?

apurvam replied:
Good catch. This would have to be called from `Sender.failBatch` and `Sender.completeBatch` as well, similar to how the `transactionManager.removeInflightBatch` is called in those methods.

yuyang08 replied:
good catch!  updated the code to call `maybeRemoveFromInflightBatches` from  `Sender.failBatch` and `Sender.completeBatch`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -333,7 +371,7 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
         // the split doesn't happen too often.
         CompressionRatioEstimator.setEstimation(bigBatch.topicPartition.topic(), compression,
-                                                Math.max(1.0f, (float) bigBatch.compressionRatio()));
+            Math.max(1.0f, (float) bigBatch.compressionRatio()));
```
nit: unintentional?

yuyang08 replied:
restored the space

hachikuji reviewed (commented):
Left a few initial comments. Still making my way through the full PR.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,37 +281,47 @@ private boolean isMuted(TopicPartition tp, long now) {
     }
 
+    private void maybeUpdateNextBatchExpiryTime(ProducerBatch batch) {
+        nextBatchExpiryTimeMs = (batch.createdMs + deliveryTimeoutMs < 0) ? nextBatchExpiryTimeMs
```
In what scenario will `batch.createdMs + deliveryTimeoutMs` be negative? `batch.createdMs` is the wall clock time when the batch is created, and should always be positive. The config def for `deliveryTimeoutMs` enforces the value is `atleast(0)`. So then how could the sum of the two be negative? Are you accounting for wrap around?

yuyang08 replied:
yes, this is to guard us against potential overflow due to setting a large value for `deliveryTimeoutMs`.

apurvam replied:
Ok. It would be good mention this in a comment. It is very non obvious.

yuyang08 replied:
update the code to use `if ...` statement and add the comment

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -483,6 +514,116 @@ public boolean hasUndrained() {
     }
 
+
+    private boolean shouldStopDrainBatchesForPartition(ProducerBatch first, TopicPartition tp) {
```
I am assuming this method and `drainBatchesForOneNode` is just a refactor? Was any logic changed here? It is hard to tell from the diff and the logic is super intricate so subtle changes are easy to miss. I assume nothing should have to change in this portion of the code since it is totally orthogonal to expiring batches.

yuyang08 replied:
yes, this is a refactoring. otherwise, the method `RecordAccumulator.drain` fails in style checking due to too many possible paths in one method.  comparing with the previous code, the change is to add the following lines on updating `inflightBatches`

                   // put this batch in the infligh list
                    List<ProducerBatch> inflightBatchList = inFlightBatches.get(batch.topicPartition);
                    if (inflightBatchList == null) {
                        inflightBatchList = new LinkedList<>();
                        inFlightBatches.put(tp, inflightBatchList);
                    }
                    inflightBatchList.add(batch);

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -691,11 +716,11 @@ private void sendProduceRequest(long now, int destination, short acks, int timeo
 
         String transactionalId = null;
-        if (transactionManager != null && transactionManager.isTransactional()) {
+        if (transactionManager != null && transactionManager.isTransactional())
```
here and elsewhere, the code style for kafka requires that there be braces even around single line if statements like this.

yuyang08 replied:
restored the curly braces.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -293,14 +315,15 @@ private long sendProducerData(long now) {
         // that isn't yet sendable (e.g. lingering, backing off). Note that this specifically does not include nodes
         // with sendable data that aren't ready to send since they would cause busy looping.
         long pollTimeout = Math.min(result.nextReadyCheckDelayMs, notReadyTimeout);
+        pollTimeout = Math.min(pollTimeout, this.accumulator.nextExpiryTimeMs() - now);
```
The preceding comment needs to be updated to account for this new logic.

yuyang08 replied:
good point. updated the comment

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -239,9 +240,30 @@ void run(long now) {
         client.poll(pollTimeout, now);
     }
 
+    private List<ProducerBatch> getExpiredBatches(Map<Integer, List<ProducerBatch>> drainedBatches,
```
This seems not to be used.

yuyang08 replied:
good catch.  removed the unused code.

apurvam reviewed (commented):
Thanks for the patch @yuyang08 .. I made a pass over the core logic and left some comments. The logic is looking good. 

I still need to go over the tests and make a second pass over the core code.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -483,6 +517,116 @@ public boolean hasUndrained() {
+                    ready.add(batch);
+
+                    // put this batch in the infligh list
+                    List<ProducerBatch> inflightBatchList = inFlightBatches.get(batch.topicPartition);
```
nit: move this to a helper named `markBatchInflight` or something similar.

yuyang08 replied:
updated the code to capture this in `markBatchInflight` method

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +281,84 @@ private boolean isMuted(TopicPartition tp, long now) {
+            // the non-negative check is to guard us against potential overflow due to setting
+            // a large value for deliveryTimeoutMs
+            nextBatchExpiryTimeMs = Math.min(nextBatchExpiryTimeMs, batch.createdMs + deliveryTimeoutMs);
+        }
```
We should log a warning if we have overflowed and are hence not updating the next batch expiry time.

yuyang08 replied:
good point. added the logging here.

apurvam commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1871,6 +1881,138 @@ private void testSplitBatchAndSend(TransactionManager txnManager,
+            request.get();
+            fail("The expired batch should throw a TimeoutException");
+        } catch (ExecutionException e) {
+            assertTrue(e.getCause() instanceof TimeoutException);
```
We should add some sort of `accumulator.hasInflightBatches` method, and then check that it returns false here. This would check that expired batches are not reenqueued, which is logic added in this patch.

yuyang08 replied:
added the method `public List<ProducerBatch> inFlightBatches(TopicPartition tp)`  in `RecordAccumulator`, and updated the test with two assertions: 

       line 1912: assertEquals("Expect one in-flight batch in accumulator", 1, accumulator.inFlightBatches(tp0).size());
        .....
       line 1920: assertEquals("Expect zero in-flight batch in accumulator", 0, accumulator.inFlightBatches(tp0).size());

apurvam commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1871,6 +1881,138 @@ private void testSplitBatchAndSend(TransactionManager txnManager,
+    }
+
+    @Test
+    public void testWhenFirstBatchExpireNoSendSecondBatchIfGuaranteeOrder() throws InterruptedException {
```
How is this different from the test `testExpiryOfFirstBatchShouldNotCauseUnresolvedSequencesIfFutureBatchesSucceed`?

yuyang08 replied:
`testExpiryOfFirstBatchShouldNotCauseUnresolvedSequencesIfFutureBatchesSucceed` initialize `sender` with `guaranteeMessageOrder = false`, while `testWhenFirstBatchExpireNoSendSecondBatchIfGuaranteeOrder` initialize sender with `guaranteeMessageOrder = true`.  The `inflightBatches` size is different when we set the parameter to true/false.

apurvam commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1871,6 +1881,138 @@ private void testSplitBatchAndSend(TransactionManager txnManager,
+        sender.run(time.milliseconds());  // expire the batch
+        assertTrue(request1.isDone());
+
+        System.out.println("client.inFlightRequestCount() = " + client.inFlightRequestCount());
```
THis should be dropped, or be `log.debug`.

yuyang08 replied:
good catch. removed this debugging line.

apurvam commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1871,6 +1881,138 @@ private void testSplitBatchAndSend(TransactionManager txnManager,
+        assertEquals(0, client.inFlightRequestCount());
+
+        sender.run(time.milliseconds()); // receive first response and do not reenqueue.
+        assertEquals(0, client.inFlightRequestCount());
```
If we had `accumulator.hasInflightRequests` we could assert false here.

apurvam reviewed (commented):
Hi @yuyang08 , thanks for the updates.

It is looking good. I left some comments, but I think the larger point is that we should add some checks that the inflight batches tracked in the accumulator are actually cleared. 

I added some suggestions for adding checks for when batches expire.

But we should also add checks that the accumulator has no inflight batches when batches complete successfully and when they fail.

apurvam commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -1871,6 +1881,138 @@ private void testSplitBatchAndSend(TransactionManager txnManager,
+        sender.run(time.milliseconds());  // expire the batch
+        assertTrue(request1.isDone());
+        // make sure the batch got deallocated right after expiry
+        assertTrue(pool.allMatch());
```
How is this testing that we don't double deallocate. Ideally, we would expire the inflight batch, then get a response, and then check that the batch is not deallocated twice. In this case, it would de allocate only once, unless I am missing something.

yuyang08 replied:
updated the test a bit to add another `sender.run(time.milliseconds());` call after  the batch expiry. Not sure if that fits `expire the inflight batch, then get a response`. 

my expectation was that if there is double deallocation, we would get an `IllegalStateException` exception from  `MatchingBufferPool.deallocate`.

yuyang08 commented:
@hachikuji, @apurvam  I've updated `SenderTest.java` with checks to ensure that the inflight batches tracked in the accumulator are cleared properly, and addressed your comments.  could you check again?

yuyang08 commented:
@hachikuji , @apurvam  friendly ping ... could you help to review the updated change to allow us iterate faster?  thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -458,10 +462,22 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
+        int requestTimeoutMs = config.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+        if (deliveryTimeoutMs < Integer.MAX_VALUE && deliveryTimeoutMs < lingerMs + requestTimeoutMs) {
+            // set deliveryTimeoutMs to lingerMs + requestTimeoutMs for backward compatibility
+            deliveryTimeoutMs = lingerMs + requestTimeoutMs;
```
To clarify, I think we should only override the default value. If the user has explicitly provided an inconsistent value, then we should throw an exception. We can check this using `config.originals()`.

yuyang08 replied:
updated

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -77,13 +77,13 @@
-    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long now, boolean isSplitBatch) {
-        this.createdMs = now;
-        this.lastAttemptMs = now;
+    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long creationTime, boolean isSplitBatch) {
```
nit: `createdMs` for consistency?

By the way, we have a function below `createdTimeMs` which is currently unused and seems incorrect anyway. Can you remove it?

yuyang08 replied:
updated to `createdMs`, and removed the unused method.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -158,34 +158,52 @@ public void abort(RuntimeException exception) {
             } else {
-                throw new IllegalStateException("Batch has already been completed in final state " + this.finalState.get());
+                // FAILED --> ABORTED and ABORTED --> FAILED transitions are ignored.
+                log.debug("Ignores state transition {} -> {} for {} batch with base offset {}",
```
nit: "Ignores" -> "Ignored"?

yuyang08 replied:
updated

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -158,34 +158,52 @@ public void abort(RuntimeException exception) {
+                    tryFinalState, topicPartition, baseOffset, this.finalState.get());
             } else {
-                throw new IllegalStateException("Batch has already been completed in final state " + this.finalState.get());
+                // FAILED --> ABORTED and ABORTED --> FAILED transitions are ignored.
```
It seems `tryFinalState` can only be SUCCEEDED or FAILED, so one of these transitions is not possible anyway.

yuyang08 replied:
`ProducerBatcn.finalState` can also be updated to `FinalState.ABORTED` through  `Sender.run() --> RecordAccumulator.abortIncompleteBatches() or abortUndrainedBatches() -->  ... -> ProducerBatch.abort()` .

hachikuji replied:
My point is that `tryFinalState` can only be SUCCEEDED or FAILED.
```java
final FinalState tryFinalState = (exception == null) ? FinalState.SUCCEEDED : FinalState.FAILED;
```
So the only possibilities are `FAILED -> FAILED` and `ABORTED -> FAILED`.

yuyang08 replied:
i see. misunderstood your comment earlier. updated the change

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -158,34 +158,52 @@ public void abort(RuntimeException exception) {
-                throw new IllegalStateException("Batch has already been completed in final state " + this.finalState.get());
+                // FAILED --> ABORTED and ABORTED --> FAILED transitions are ignored.
+                log.debug("Ignores state transition {} -> {} for {} batch with base offset {}",
+                    tryFinalState, this.finalState.get(), topicPartition, baseOffset);
```
Are the order of the first two arguments backwards? I think `tryFinalState` is the state we're trying to transition to.

yuyang08 replied:
good catch!  fixed the arguments order.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -158,34 +158,52 @@ public void abort(RuntimeException exception) {
             log.trace("Successfully produced messages to {} with base offset {}.", topicPartition, baseOffset);
-            finalState = FinalState.SUCCEEDED;
         } else {
             log.trace("Failed to produce messages to {}.", topicPartition, exception);
```
nit: Can we add `baseOffset` to this message?

yuyang08 replied:
updated the log message to include baseOffset

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -299,30 +317,24 @@ public String toString() {
     }
 
+    boolean hasReachedDeliveryTimeout(long deliveryTimeoutMs, long now) {
+        return  deliveryTimeoutMs <= now - this.createdMs;
```
nit: unneeded space after `return`

yuyang08 replied:
good catch. removed the redundant whitespace

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -299,30 +317,24 @@ public String toString() {
 
     /**
-     * If {@link #maybeExpire(int, long, long, long, boolean)} returned true, the sender will fail the batch with
+     * If {@link #hasReachedDeliveryTimeout(long, long)} returned true, the sender will fail the batch with
```
I'm finding this API confusing. The caller is first required to check `hasDeliveryTimeout`. If true, then they should call `expire()` so that `expiryErrorMessage` gets set. Then they need to call `timeoutException()` and pass the result to `done()`. This is so tangled with with the producer's internal logic that there's almost no point in having it as a separate class.

I think we can just get rid of `timeoutException` and `expire`. Instead of calling `expire`, the caller can invoke `abortRecordAppends` directly. For the one usage of `timeoutException` in `Sender`, we can construct the `TimeoutException` object directly.

yuyang08 replied:
good point. removed `expire` and `timeoutException` methods from `ProducerBatch`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -122,14 +130,16 @@ public RecordAccumulator(LogContext logContext,
         this.lingerMs = lingerMs;
         this.retryBackoffMs = retryBackoffMs;
+        this.deliveryTimeoutMs = deliveryTimeoutMs;
+        this.nextBatchExpiryTimeMs = Long.MAX_VALUE;
```
nit: this is already initialized above

yuyang08 replied:
removed this line

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +281,87 @@ private boolean isMuted(TopicPartition tp, long now) {
+            // a large value for deliveryTimeoutMs
+            nextBatchExpiryTimeMs = Math.min(nextBatchExpiryTimeMs, batch.createdMs + deliveryTimeoutMs);
+        } else {
+            log.warn("Skipping next batch expiry time update due to addition overflow: "
```
This will be very spammy if the user has set a large delivery timeout. I don't think the log message has much value to be honest, but perhaps we should at least reduce the level to TRACE?

yuyang08 replied:
`batch.createdMs + deliveryTimeoutMs ` should be positive.  we log an warning here in case of addition overflow. we have defined `deliverTimeoutMs` as `int` when we initialize the producer. Given that, i am not expecting that that log line will be trigged often. 

       private static int configureDeliveryTimeout(ProducerConfig config, Logger log) {
            int deliveryTimeoutMs = config.getInt(ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG);
            int lingerMs = config.getInt(ProducerConfig.LINGER_MS_CONFIG);
            int requestTimeoutMs = config.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);

Currently I keep RecordAccumulator constructor signature unchanged. If we change the type of 'deliveryTimeoutMs', it is better to also change the type of `lingerMs` and `retryBackoffMs` to be consistent. As changing the types of those paramters will require changes in many test cases to pass `int` instead of `long`,  I'd rather to put the RecordAccumulator constructor signature change in other PR. 


        public RecordAccumulator(LogContext logContext,
                                 int batchSize,
                                 CompressionType compression,
                                 long lingerMs,
                                 long retryBackoffMs,
                                 long deliveryTimeoutMs,
                                  ...
.

hachikuji replied:
Fair enough. I missed the long coercion. I think the log message has dubious value, but I agree we shouldn't ever actually see it.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +281,87 @@ private boolean isMuted(TopicPartition tp, long now) {
+            List<ProducerBatch> partitionInFlightBatches = inFlightBatches.get(entry.getKey());
+            if (partitionInFlightBatches != null) {
+                Iterator<ProducerBatch> iter = partitionInFlightBatches.iterator();
+                while (iter.hasNext()) {
```
If the list is empty after this iteration, should we remove the partition from `inFlightBatches`?

yuyang08 replied:
added logic on removing the partition from `inFlightBatches` if the corresponding `ProduceBatch` list is empty.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +281,87 @@ private boolean isMuted(TopicPartition tp, long now) {
-            // This is to prevent later batches from being expired while an earlier batch is still in progress.
-            // Note that `muted` is only ever populated if `max.in.flight.request.per.connection=1` so this protection
-            // is only active in this case. Otherwise the expiration order is not guaranteed.
-            if (!isMuted(tp, now)) {
```
We seem to have lost this check as well as the synchronization below. Are they no longer needed?

yuyang08 replied:
delivery.timeout.ms enforces the time limit in batching in the accumulator, retries, and in the inflight segment, we should examine all batches on expiration, no matter the partition is muted or not.  given that, we no longer need the `isMuted` check. 

If we added `isMuted`, there will be test failures in `RecordAccumulatorTest`: testExpiredBatchSingle, testExpiredBatchSingleMaxValue, testExpiredBatches etc. Fixing those tests by considering whether a partition is muted or not complicates the logic. It is better to keep the implementation simple.

yuyang08 replied:
adding  `synchronized(deque)` back. it is still needed to guarantee mutual exclusive access to `deque` between `Sender` and `RecordAccumulator.append`.

apurvam replied:
Why did those tests fail if you leave the `isMuted` in there?

yuyang08 replied:
In `Sender.sendProducerData`,  `RecordAccumulator.drain` is called before  `accumulator.expiredBatches`.   Keeping `isMuted(....)` will have us skip muted topic partitions while checking expired batches. If the batch expires before `RecordAccmulator.drain` is called again,   in the next `Sender.sendProducerData` call, the batch will get drained and be sent to the cluster. This violates the contract that  batches that reaches delivery timeout should not be sent out.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +281,87 @@ private boolean isMuted(TopicPartition tp, long now) {
+                    ProducerBatch batch = iter.next();
+                    if (batch.hasReachedDeliveryTimeout(deliveryTimeoutMs, now)) {
+                        iter.remove();
+                        if (batch.finalState() == null) {
```
Is it possible for `finalState` to be non-null at this point now that we remove from in-flight batches in `completeBatch` and `failBatch`?

yuyang08 replied:
good question. `RecordAcumulator.expireBatches` is called from `Sender.run --> Sender.sendProducerData`, before `client.poll`. The batch state update only happens through `client.poll`. Given that,  `batch.finalState()` should always be `null` at this point.  To be cautious, I still would like to keep the `if` statement. I added an `else` statement to log an warning if we see `batch.finalstate != null`.  in theory that statement should never be executed.  Also added comments before the if statement to explain the motivation. 

                           } else {
                                log.warn("{} batch crated at {} gets unexpected final state : {}",
                                   batch.topicPartition, batch.createdMs, batch.finalState());
                           }

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -483,6 +523,119 @@ public boolean hasUndrained() {
+        return ready;
+    }
+
+    private void markBatchInflight(ProducerBatch batch) {
```
nit: maybe `addToInFlightBatches` for consistency with `maybeRemoveFromInflightBatches`

yuyang08 replied:
updated the method name to `addToInFlightBatches `

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -784,4 +856,4 @@ public ReadyCheckResult(Set<Node> readyNodes, long nextReadyCheckDelayMs, Set<St
     }
 
-}
+}
```
nit: add back the newline

yuyang08 replied:
added it back

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -239,9 +239,9 @@ void run(long now) {
         client.poll(pollTimeout, now);
     }
 
+
```
nit: unneeded newline

yuyang08 replied:
removed the new line

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -564,7 +560,8 @@ else if (error == Errors.CLUSTER_AUTHORIZATION_FAILED)
                 if (error.exception() instanceof UnknownTopicOrPartitionException) {
                     log.warn("Received unknown topic or partition error in produce request on partition {}. The " +
-                            "topic/partition may not exist or the user may not have Describe access to it", batch.topicPartition);
+                            "topic/partition may not exist or the user may not have Describe access to it",
```
nit: `topic-partition` as we did above?

yuyang08 replied:
updated

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -589,29 +586,36 @@ private void reenqueueBatch(ProducerBatch batch, long currentTimeMs) {
 
-    private void failBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response, RuntimeException exception, boolean adjustSequenceNumbers) {
+    private void failBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response, RuntimeException exception,
+        boolean adjustSequenceNumbers) {
```
nit: realign

yuyang08 replied:
updated

hachikuji commented on clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:
```diff
@@ -845,9 +843,9 @@ else if (value instanceof Class)
         private final Number max;
 
-        private Range(Number min, Number max) {
+        private Range(Number min, Number maxInclusive) {
```
Aren't both bounds inclusive? The inconsistency is a little annoying and this is a private constructor anyway. I think mentioning this in the javadoc below is good enough.

yuyang08 replied:
added java doc comments and changed the parameter name to `max`

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

hachikuji commented on docs/upgrade.html:
```diff
@@ -30,6 +30,7 @@ <h4><a id="upgrade_2_1_0" href="#upgrade_2_1_0">Upgrading from 0.8.x, 0.9.x, 0.1
         offset retention period (or the one set by broker) has passed since their last commit.</li>
     <li>The default for console consumer's <code>enable.auto.commit</code> property when no <code>group.id</code> is provided is now set to <code>false</code>.
         This is to avoid polluting the consumer coordinator cache as the auto-generated group is not likely to be used by other consumers.</li>
+    <li>The default value for <code>retries</code> was changed to <code>INT.MAX_VALUE</code>, as we introduces <code>delivery.timeout.ms</code> that sets the upper bound of message deliver timeout. </li>
```
Can you mention that this is for the producer? For example:

>  The default value for the producer's <code>retries</code> config was changed to <code>INT.MAX_VALUE</code> ...

Might also be worthwhile including a link to KIP-91

yuyang08 replied:
updated to include KIP-91 link

yuyang08 commented:
@hachikuji  addressed all of your comments. could you take a look again?

yuyang08 commented:
@hachikuji @apurvam  @ijuma  have updated the change to address your comments. could you take a look again? thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +280,96 @@ private boolean isMuted(TopicPartition tp, long now) {
+                    deque.addFirst(batch);
+            }
+        } else {
+            batch.abortRecordAppends();
```
I'm not sure this is sufficient to solve the issue mentioned above. If we do not reenqueue because the timeout has been reached, then who is responsible for completing the batch? I think I would probably suggest that we skip the check for delivery timeout and just reenqueue. We will detect the expiration the next time we iterate the deque.

apurvam replied:
I think the only issue with reenqueing unconditionally is that the batch will then be drained even if it is expired, since `accumulator.drain` is called before `accumulator.expiredBatches` in the background thread. This could violate the contract.

That said, we need to complete the batch and deallocate it over here, otherwise it seems to be dropped on the flor.

yuyang08 replied:
update to check whether a batch has reached `deliveryTimeoutMs` or not in `Sender.canRetry`.  With this change, we do not need to check whether a batch has timeout or not in `RecordAccumuator.reenqueue`, and cleaning up of timed-out batches is handled by `Sender.failBatch`. 

       private boolean canRetry(ProducerBatch batch, ProduceResponse.PartitionResponse response, long now) {
            return !batch.hasReachedDeliveryTimeout(accumulator.getDeliveryTimeoutMs(), now) 
                    && ...

yuyang08 commented:
@hachikuji , @apurvam  friendly ping... mind to take another look?  thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +280,96 @@ private boolean isMuted(TopicPartition tp, long now) {
+                        // The batch.finalState() == null invariant should always hold. We put an if ... else...
+                        // statement here to log info of the batches that violate the invariant.
+                        if (batch.finalState() == null) {
+                            batch.abortRecordAppends();
```
Is this call necessary for in-flight batches? Presumably we have already closed the append stream since the batch was already sent once.

yuyang08 replied:
good point. removed this line.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +280,96 @@ private boolean isMuted(TopicPartition tp, long now) {
+                        iter.remove();
+                        // expireBatches is called in Sender.sendProducerData, before client.poll.
+                        // The batch.finalState() == null invariant should always hold. We put an if ... else...
+                        // statement here to log info of the batches that violate the invariant.
```
If it is an expected invariant, I would suggest we raise an exception if it does not hold. Otherwise, bugs will go undetected since we don't have a way to verify logging output when running tests.

apurvam replied:
+1. Better to fail with an exception if the invariant is violated.

yuyang08 replied:
updated the change to throw `IllegalBatchFinalStateException` when the invariant is violated.

hachikuji commented on docs/upgrade.html:
```diff
@@ -30,6 +30,10 @@ <h4><a id="upgrade_2_1_0" href="#upgrade_2_1_0">Upgrading from 0.8.x, 0.9.x, 0.1
         offset retention period (or the one set by broker) has passed since their last commit.</li>
     <li>The default for console consumer's <code>enable.auto.commit</code> property when no <code>group.id</code> is provided is now set to <code>false</code>.
         This is to avoid polluting the consumer coordinator cache as the auto-generated group is not likely to be used by other consumers.</li>
+    <li>The default value for <code>retries</code> was changed to <code>INT.MAX_VALUE</code>, as we introduces <code>delivery.timeout.ms</code>
```
nit: introduces -> introduced

yuyang08 replied:
fixed the typo.

hachikuji reviewed (commented):
Thanks, left a few more comments.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -391,18 +392,21 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
-                    this.totalMemorySize,
                     this.compressionType,
-                    config.getLong(ProducerConfig.LINGER_MS_CONFIG),
+                    config.getInt(ProducerConfig.LINGER_MS_CONFIG),
```
Why changing linger to int ?

yuyang08 replied:
we have `request.timeout.ms` and `delivery.timeout.ms` of `int` type. this is to make the type of `linger.ms` be consistent with other timeout related settings.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -458,10 +462,30 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
 
-    private static TransactionManager configureTransactionState(ProducerConfig config, LogContext logContext, Logger log) {
+    private static int configureDeliveryTimeout(ProducerConfig config, Logger log) {
+        int deliveryTimeoutMs = config.getInt(ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG);
```
Should this be of type long ?
With long, there is no overflow on line 478

In ProducerBatch, deliveryTimeoutMs is long in hasReachedDeliveryTimeout

yuyang08 replied:
`ProducerBatch.hasReachedDeliveryTimeout` is called by RecordAccumulator. In RecordAccumulator's construct, we have had using `long` type for `lingerMs`, and `retryBackoffMs`. It will be inconsistent to use `int` for `deliveryTimeoutMs`. And it will require changes at many places (especially in the test cases) if we use `int` type for `lingerMs` and `retryBackoffMs`.  I thought that it would be better to have another PR for data type related changes for `lingerMs` etc.

        public RecordAccumulator(LogContext logContext,
                                 int batchSize,
                                 CompressionType compression,
                                 long lingerMs,
                                 long retryBackoffMs,
                                 ...

tedyu commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -458,10 +462,30 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
+            } else {
+                // override deliveryTimeoutMs default value to lingerMs + requestTimeoutMs for backward compatibility
+                deliveryTimeoutMs = lingerMs + requestTimeoutMs;
+                log.warn("{} should be equal to or larger than {} + {}. Setting it to {}.",
```
Doesn't need to be warn.
Can be info since there is no action from user

yuyang08 replied:
This is try to get the user's attention as we are overriding the default  delivery.timeout.ms setting.  Previously the user may set a long `request.timeout.ms` as a work around. The user may want to explicitly set `delivery.timeout.ms` and give a smaller value for `request.timeout.ms`.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -85,12 +87,16 @@
     private final TransactionManager transactionManager;
 
+    // A per-partition queue of batches ordered by creation time for tracking the in-flight batches
+    private final ConcurrentMap<TopicPartition, List<ProducerBatch>> inFlightBatches;
```
It seems to me that we need to synchronize on accesses to the contained `List<ProducerBatch>`. This list is updated in `maybeRemoveFromInflightBatches`, which called both from the sender background thread and from the network client threads. This unsafe access can corrupt the list in the present patch.

yuyang08 replied:
I am not fully convinced that there is concurrent access to `List<ProducerBatch> inFlightBatches`.  In KafkaProducer, there is only one background I/O thread (`Sender`) that turns the records into requests and transmit them to the cluster.  `inFlightBatches` is only being modified by `RecordAccumulator` methods that are called by `Sender` thread.   `Sender.run(long now)` --> `NetworkClient.poll` --> `NetworkClient.handle...` --> `Sender.handleProduceResponse` -->  `Sender.completeBatch` --> `Sender.reenqueueBatch` --> `RecordAccumulator.reenque` --> `RecordAccumulator.maybeRemoveFromInflightBatches`  is the call stack.   `NetworkClient` uses `Selector` for non-blocking i/o multiplexing, but it is not a thread. 

We need to synchronize access on `Deque<ProducerBatch> deque = entry.getValue();` as  both the producer thread and the sender thread can concurrently access `deque` at the same time.

hachikuji replied:
I couldn't find the concurrent access either, so it may be ok. It actually makes me wonder why we are using a `ConcurrentMap`? To be honest, it would be more intuitive for the inflight batches to be tracked inside `Sender`. I'd suggest doing that if it's straightforward, but I'm fine leaving it here if it takes a lot of work.

apurvam replied:
Yea. I think I was mistaken. If the callbacks from the network client are happening in the context of the sender thread, then there is no concurrent access on `inflightBatches`.

yuyang08 replied:
good point on tracking the inflight batches in `Sender`.  updated the change to move `inFlightBatches` to `Sender`,  and changed its type from `ConcurrentMap` to `Map`.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +280,96 @@ private boolean isMuted(TopicPartition tp, long now) {
                         } else {
-                            // Stop at the first batch that has not expired.
-                            break;
+                            log.warn("{} batch crated at {} gets unexpected final state : {}",
```
typo: "created".

yuyang08 replied:
fixed the typo

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,55 +280,96 @@ private boolean isMuted(TopicPartition tp, long now) {
+            TopicPartition topicPartition = entry.getKey();
+            List<ProducerBatch> partitionInFlightBatches = inFlightBatches.get(topicPartition);
+            if (partitionInFlightBatches != null) {
+                Iterator<ProducerBatch> iter = partitionInFlightBatches.iterator();
```
As mentioned above, we should synchronize on this list while iterating, and then synchronize again in `maybeRemoveFromInflightBatches`

yuyang08 replied:
As I mentioned earlier, i might miss something, but it seems to me that there is no concurrent access to `inFlightBatches`, and we do not need synchronize qualifier for it.

apurvam reviewed (commented):
Sorry for the delayed response @yuyang08 . The week of july 4 was extremely short, and last week we had a company event and multi-day off site, which took a lot of cycles.

I left a few more comments, the biggest one around synchronization of `inFlightBatches`.

asfgit commented:
FAILURE
 4401 tests run, 1 skipped, 1 failed.
--none--

yuyang08 commented:
@hachikuji , @apurvam , @tedyu  i updated the change and addressed all of your comments. could you take a look again?  thanks!

yuyang08 commented:
@hachikuji , @apurvam, @tedyu  thanks for your review!  I've address all of your comments. The change also passed the tests. could you take a look again?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
+                        if (batch.finalState() == null) {
+                            expiredBatches.add(batch);
+                        } else {
+                            throw new IllegalBatchFinalStateException(batch.topicPartition + " batch created at " +
```
I'm not sure we need a new exception type for an error that should not happen. Maybe we can just use `IllegalStateException`?

yuyang08 replied:
updated to use `IllegalStateException`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
+        for (ProducerBatch batch : batches) {
+            List<ProducerBatch> inflightBatchList = inFlightBatches.get(batch.topicPartition);
+            if (inflightBatchList == null) {
+                inflightBatchList = new LinkedList<>();
```
nit: can we use `ArrayList`?  https://twitter.com/joshbloch/status/583813919019573248

yuyang08 replied:
updated to use ArrayList

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
+    }
+
+    public void addToInflightBatches(Map<Integer, List<ProducerBatch>> batches) {
+        for (Map.Entry<Integer, List<ProducerBatch>> entry : batches.entrySet()) {
```
We're not using the keys, so can we just iterate the values?

yuyang08 replied:
update to iterate through `batches.values()`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -272,46 +272,53 @@ private boolean isMuted(TopicPartition tp, long now) {
+        }
+    }
+
+    public Set<TopicPartition>  getTopicPartitions() {
```
nit: extra space before method name

yuyang08 replied:
removed the extra space

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
+     *  Get the in-flight batches that has reached delivery timeout.
+     */
+    private List<ProducerBatch> getExpiredInflightBatches(long now) {
+        Set<TopicPartition>  topicPartitions = this.accumulator.getTopicPartitions();
```
Can you clarify why use this for the iteration? I would have expected we would iterate over the entry set of `inFlightBatches`.

yuyang08 replied:
this is really not needed.  updated to iterate through `inFlightBatches` directly.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
+        this.inFlightBatches = new HashMap<>();
+    }
+
+    public Map<TopicPartition, List<ProducerBatch>> getInflightBatches() {
```
nit: this is unused

yuyang08 replied:
removed this unused method

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
+                        iter.remove();
+                        // expireBatches is called in Sender.sendProducerData, before client.poll.
+                        // The batch.finalState() == null invariant should always hold. We put an if ... else...
+                        // statement here to log info of the batches that violate the invariant.
```
This message needs to be updated.

yuyang08 replied:
updated the comment: 

                        // expireBatches is called in Sender.sendProducerData, before client.poll.
                        // The batch.finalState() == null invariant should always hold. An IllegalStateException
                        // exception will be thrown if the invariant is violated.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
+
+    public void maybeRemoveFromInflightBatches(ProducerBatch batch) {
+        List<ProducerBatch> batches = inFlightBatches.get(batch.topicPartition);
+        if (batches != null) {
```
If `batches` is empty, should we remove the list from `inFlightBatches`?

yuyang08 replied:
updated the change to remove the list from `inFlightBatches` when it is empty.

hachikuji reviewed (commented):
@yuyang08 Thanks for the updates. I think we're pretty close. I had a few more small comments.

hachikuji commented:
retest this please

apurvam reviewed (approved):
Thanks for the patch @yuyang08 and for the patience during the review!

I went over the core flow and the tests once more, and it has been simplified quite nicely and LGTM!

hachikuji committed:
Remove unused method

hachikuji committed:
Additional detail in upgrade notes

hachikuji reviewed (approved):
LGTM, merging to trunk. Thanks for the patch!

hachikuji commented:
Note that I removed an unused method and made some tweaks to the upgrade notes.

ijuma commented:
@yuyang08 Thanks for picking up this important improvement. And thanks @sutambe for submitting the original PR.

likan999 commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -459,10 +463,30 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
+        int lingerMs = config.getInt(ProducerConfig.LINGER_MS_CONFIG);
+        int requestTimeoutMs = config.getInt(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG);
+
+        if (deliveryTimeoutMs < Integer.MAX_VALUE && deliveryTimeoutMs < lingerMs + requestTimeoutMs) {
```
(deliveryTimeoutMs < lingerMs + requestTimeoutMs) implies (deliveryTimeoutMs < Integer.MAX_VALUE), why do we need to check (deliveryTimeoutMs < Integer.MAX_VALUE), logically it can be removed.
